### PR TITLE
fix: mark model as registered only if model is registered successfully

### PR DESCRIPTION
**Pull request by anyin233**
This may be one of the solution of issue #76 by move 

```python
self.registered_models[model_name] = model_config
self.request_routers[model_name] = request_router
```

to the end of SllmController.register

### CODE REVIEW

**Comment by future-xy** [comment]
This fix looks good to me. Could you provide a unit test for controller to verify such behaviour?

File: serverless_llm/serve/app_lib.py [hunk 1]
```diff
@@ -50,7 +50,10 @@ async def register_handler(request: Request):
             raise HTTPException(
                 status_code=500, detail="Controller not initialized"
             )
-        await controller.register.remote(body)
+        try:
+            await controller.register.remote(body)
+        except Exception as e:
+            raise HTTPException(status_code=500, detail="Cannot register model, please contant the administrator")
```
Comment (JustinTong0323) [inline serverless_llm/serve/app_lib.py:56]: typo here, "contant" should be "contact"?

**Review by JustinTong0323 — CHANGES_REQUESTED** [review]

File: serverless_llm/serve/app_lib.py (hunk 1 above)
Comment (anyin233) [reply serverless_llm/serve/app_lib.py]: Fixed

**Comment by anyin233** [comment]
I added a unit simple test for `SllmController.register()` by modifying some code in the controller to remove Ray during unit testing.

**Review by future-xy — CHANGES_REQUESTED** [review]

**Review by future-xy — CHANGES_REQUESTED** [review]
The current code doesn't seem to be working in my tests. Below is the error message:
```
(sllm) fy@fuji2:~/ServerlessLLM$ pytest tests/controller_test/
/home/fy/miniconda3/envs/sllm/lib/python3.10/site-packages/pytest_asyncio/plugin.py:208: PytestDeprecationWarning: The configuration option "asyncio_default_fixture_loop_scope" is unset.
The event loop scope for asynchronous fixtures will default to the fixture caching scope. Future versions of pytest-asyncio will default the loop scope for asynchronous fixtures to function scope. Set the default fixture loop scope explicitly in order to avoid unexpected behavior in the future. Valid fixture loop scopes are: "function", "class", "module", "package", "session"

  warnings.warn(PytestDeprecationWarning(_DEFAULT_FIXTURE_LOOP_SCOPE_UNSET))
========================================================================== test session starts ==========================================================================
platform linux -- Python 3.10.14, pytest-8.3.3, pluggy-1.5.0
rootdir: /home/fy/ServerlessLLM
configfile: pyproject.toml
plugins: anyio-4.6.0, asyncio-0.24.0
asyncio: mode=strict, default_loop_scope=None
collected 1 item                                                                                                                                                        

tests/controller_test/controller_test.py E                                                                                                                        [100%]

================================================================================ ERRORS =================================================================================
_________________________________________________________________ ERROR at setup of test_network_failed _________________________________________________________________

setup_environment = <ray.cluster_utils.Cluster object at 0x7fb8dbe06200>

    @pytest.fixture(scope="session")
    def setup_controller(setup_environment):
        cluster = setup_environment
        cluster.add_node(num_cpus=4, num_gpus=1, object_store_memory=75 *
                         1024 * 1024, resources={"worker_node": 1, "worker_id_0": 1})
>       controller = SllmController({"hardware_config": None}, debug=True)
E       TypeError: SllmController.__init__() got an unexpected keyword argument 'debug'

tests/controller_test/controller_test.py:48: TypeError
------------------------------------------------------------------------- Captured stderr setup -------------------------------------------------------------------------
2024-09-26 13:39:18,842 INFO utils.py:112 -- Overwriting previous Ray address (129.215.164.142:6379). Running ray.init() on this node will now connect to the new instance at 129.215.164.142:56129. To override this behavior, pass address=129.215.164.142:6379 to ray.init().
2024-09-26 13:39:18,843 INFO worker.py:1540 -- Connecting to existing Ray cluster at address: 129.215.164.142:56129...
2024-09-26 13:39:18,847 INFO worker.py:1724 -- Connected to Ray cluster.
======================================================================== short test summary info ========================================================================
ERROR tests/controller_test/controller_test.py::test_network_failed - TypeError: SllmController.__init__() got an unexpected keyword argument 'debug'
=========================================================================== 1 error in 3.74s ============================================================================
(sllm) fy@fuji2:~/ServerlessLLM$ 
```

If unit testing with Ray proves too challenging, we can consider pausing it for now.

**Comment by anyin233** [comment]
> The current code doesn't seem to be working in my tests. Below is the error message:
> 
> ```
> (sllm) fy@fuji2:~/ServerlessLLM$ pytest tests/controller_test/
> /home/fy/miniconda3/envs/sllm/lib/python3.10/site-packages/pytest_asyncio/plugin.py:208: PytestDeprecationWarning: The configuration option "asyncio_default_fixture_loop_scope" is unset.
> The event loop scope for asynchronous fixtures will default to the fixture caching scope. Future versions of pytest-asyncio will default the loop scope for asynchronous fixtures to function scope. Set the default fixture loop scope explicitly in order to avoid unexpected behavior in the future. Valid fixture loop scopes are: "function", "class", "module", "package", "session"
> 
>   warnings.warn(PytestDeprecationWarning(_DEFAULT_FIXTURE_LOOP_SCOPE_UNSET))
> ========================================================================== test session starts ==========================================================================
> platform linux -- Python 3.10.14, pytest-8.3.3, pluggy-1.5.0
> rootdir: /home/fy/ServerlessLLM
> configfile: pyproject.toml
> plugins: anyio-4.6.0, asyncio-0.24.0
> asyncio: mode=strict, default_loop_scope=None
> collected 1 item                                                                                                                                                        
> 
> tests/controller_test/controller_test.py E                                                                                                                        [100%]
> 
> ================================================================================ ERRORS =================================================================================
> _________________________________________________________________ ERROR at setup of test_network_failed _________________________________________________________________
> 
> setup_environment = <ray.cluster_utils.Cluster object at 0x7fb8dbe06200>
> 
>     @pytest.fixture(scope="session")
>     def setup_controller(setup_environment):
>         cluster = setup_environment
>         cluster.add_node(num_cpus=4, num_gpus=1, object_store_memory=75 *
>                          1024 * 1024, resources={"worker_node": 1, "worker_id_0": 1})
> >       controller = SllmController({"hardware_config": None}, debug=True)
> E       TypeError: SllmController.__init__() got an unexpected keyword argument 'debug'
> 
> tests/controller_test/controller_test.py:48: TypeError
> ------------------------------------------------------------------------- Captured stderr setup -------------------------------------------------------------------------
> 2024-09-26 13:39:18,842 INFO utils.py:112 -- Overwriting previous Ray address (129.215.164.142:6379). Running ray.init() on this node will now connect to the new instance at 129.215.164.142:56129. To override this behavior, pass address=129.215.164.142:6379 to ray.init().
> 2024-09-26 13:39:18,843 INFO worker.py:1540 -- Connecting to existing Ray cluster at address: 129.215.164.142:56129...
> 2024-09-26 13:39:18,847 INFO worker.py:1724 -- Connected to Ray cluster.
> ======================================================================== short test summary info ========================================================================
> ERROR tests/controller_test/controller_test.py::test_network_failed - TypeError: SllmController.__init__() got an unexpected keyword argument 'debug'
> =========================================================================== 1 error in 3.74s ============================================================================
> (sllm) fy@fuji2:~/ServerlessLLM$ 
> ```
> 
> If unit testing with Ray proves too challenging, we can consider pausing it for now.

Actually, unit testing with Ray is quite difficult because Mock may not behave as expected in some scenarios. In the future, unit tests might need to launch a real server.

**Comment by anyin233** [comment]
I have removed unit test for controller.

**Review by future-xy — APPROVED** [review]

**Review by JustinTong0323 — APPROVED** [review]
LGTM

**Pull request merged**